### ISSUE TITLE
Feature resolving

### ISSUE THREAD

**Issue by Cyrik** [jcpsantiago/bulgogi#6]
Currently, features are strings that get looked up as function names in a single namespace.
This makes it easy to work with as long as you have a small feature set, but brittle in regards to mistyping. The feature namespace is also going to get pretty large potentially.

I'd like to brainstorm about solutions a little:

### namespaced keywords

- a little more complicated than the strings
- user has to know the namespace
- IDEs can help navigate, check for misspellings
- trivial to use multiple namespaces

### keep strings or non-namespaced keywords, but lookup in multiple namespaces

- simple to use
- hard to find the implementation for the user
- how do we know which namespaces to check?
- naming conflicts

### keep strings or non-namespaced keywords, but force registration of the feature functions

- simple to use
- system can be asked about all available features
- more state in the system
- feature creators have to know how to register a feature
- naming conflicts

any thoughts?

**Comment by jcpsantiago** [comment]
Thanks for keeping bringing ideas to this :)

The original intention behind using strings was to have the simplest interface possible for JSON requests. The goal is to enable any client to have as little knowledge as possible about the inner workings of bulgogi. This requirement comes from my experience deploying models to prod: I save an array of needed features as `["feature1", "feature2"]` in some kv-store and at runtime some mechanism which decides which model to use (let's call it a "model router") decides which model to use, fetches which features are needed, sends the req to bulgogi, collects the response and applies the model to the data. Example flow (some steps missing ofc):
`get training data from some store (features/cols match names in bulgogi, because they were saved by bulgogi itself) -> train model ->store features needed in kv store-> "model router" picks model and  features-> send to bulgogi-> send to model`

This also allows for a soft-coupling between bulgogi and any client: the only thing the clients need to know is a name, and whoever maintains bulgogi is free to move things around as they see fit.

Out of your options, the one I gave the most thought so far is option 2, but it would need some sort of build-time procedure to prevent deployments containing features with the same name across namespaces. Effectively, it's not possible to _reuse names_ at all, but it's still possible to have multiple namespaces e.g. one could read all namespaces `project.features.xxx` when starting. This would still allow for the organisation of features beyond a single namespace e.g. using "thematic" ns. Not super great though, because of the external build-time dependency on a check before deployment. If Bulgogi is "just" a library to coordinate the feature calculation that is out of our hands, and will likely lead to tears when someone shoots themselves in the foot.

Registering is useful, and I would even say it's something we _always_ want in a real prod env because we can also implement checks for model deployment such as "only deploy model if needed features are in prod", otherwise models go to production and explode due to lack of data. Not fun, so I think most people will (should) want to do this. For this I thought about a macro `def-feature` wrapping `defn`, so feature creators would know this is the way to do it and it would feel like a technicality. No idea how to handle the naming conflicts though (besides CI, like above).

So, in summary, I think options 2 and 3 are more aligned with the philosophy I had in mind for this, but I don't know how to deal with the naming conflicts.

**Comment by iGEL** [comment]
@jcpsantiago I think, it would be great to have features be "registered" by implementing the function in a namespace with a certain prefix, eg. `bulgogi.features.*` This way, you make sure you don't have a miss match between the registry and the implementation.

To get the available features, something like this might work:

```clojure
(->> (all-ns)
     (map ns-name)
     (filter #(clojure.string/starts-with? (name %) "bulgogi.features."))
     (map ns-publics)
     (mapcat keys))
```

To detect conflicts, we could just use `frequencies`:
```clojure
(->> features
     frequencies
     (filter #(< 1 (last %)))
     (map first))
```

If that list is not empty, we have a conflict.

**Comment by Cyrik** [comment]
@iGEL using ns names is a little brittle, but I was thinking we could use namespace metadata:
```clojure
(ns foo.core
  {:bulgogi/features true}
  (:require [bar.core :as bar]))
```
What do you think?

I'll play around with a few different ways that could detect conflicts at compile time.

@jcpsantiago by model, do you mean something like?
```clojure
{:input-data {:current-amount 700
	      :email "squadron42@starfleet.ufp"
 :features ["n-digits-in-email-name" 
	    "contains-risky-item"]}
```

**Comment by jcpsantiago** [comment]
Thanks for putting it here @iGEL, I appreciate it.

@Cyrik so in your example the metadata is signalling "this namespace has features"? That would then be a nice addon to @iGEL's proposal, not a completely different approach, right? e.g. filter based on the metadata, but still do the rest of the steps with frequencies and so on to make sure there are duplicates.

In my long message above when I say _model_, I mean a _machine learning model_ (ML), not a _data_ model if that's what you were asking about. So the map you posted is an example of a payload for bulgogi and it could come from a data scientist wanting to get some data for an analysis, or it could be machine learning model in production asking for the transformed data it needs (I'm being a bit handwavy here, it's not 100% like this). Makes sense? ideally Bulgogi does not know anything about the outside world or the concept of ML models.

**Comment by Cyrik** [comment]
@jcpsantiago yes it would work the same way @iGEL mentioned. I'm just still trying to find the most robust way to report name collisions at compile time.

> In my long message above when I say model, I mean a machine learning model (ML)

That makes more sense, thank you for the explanation. But then I'm wondering how you would want to accomplish "only deploy model if needed features are in prod"? Unless you only need a list of available features in a text file or something, that should be easy to produce.

**Comment by jcpsantiago** [comment]
Presently, using our feature store at Billie ([blog post here](https://medium.com/billie-finanzratgeber/unlocking-our-data-with-a-feature-store-402ade0743b)), we only have batch features and there's a db table with all the available features. During model training, I get this list from the database and

1. do not allow the model to be trained, unless it only uses features present in that list
2. that list is saved together with the model (model in one file, features needed in a json file)
3. during "build" in CI, we check again if the features inside that json match what we see in the db and allow/block deployment as necessary

For Bulgogi we can do the same and/or (@iGEL's suggestion) have a endpoint which returns all available features and do the same procedure as above.

It was natural for me to develop it like this, but it might not be for everyone. In any case, having some sort of endpoint to get the features sounds like a no-brainer to me.

**Comment by Cyrik** [comment]
Yes, makes total sense. I just wanted to be sure that we hit your use case. Reporting all features is going to be easy in all of those versions.

**Comment by jcpsantiago** [comment]
Since you contributed an initial solution for this in https://github.com/jcpsantiago/bulgogi/pull/4, can we close this issue, or are there things you still would like to discuss under this topic?

**Comment by Cyrik** [comment]
No, I think I'm happy to close it, thanks!
